electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +263,112 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
+    control_voltage = self.btn_in.link().voltage.hull(self.gnd.link().voltage)
+    pwr_voltage = self.pwr_out.link().voltage.hull(self.gnd.link().voltage)
+    pwr_current = self.pwr_out.link().current_drawn.hull(RangeExpr.ZERO)
+
+    self.pull_res = self.Block(Resistor(
+      resistance=self.pull_resistance
+    ))
+    self.pwr_fet = self.Block(Fet.PFet(
+      drain_voltage=pwr_voltage,
+      drain_current=pwr_current,
+      gate_voltage=(-control_voltage.upper(), control_voltage.upper()),  # TODO this ignores the diode drop
+    ))
+
+    self.amp_res = self.Block(Resistor(
+      resistance=self.amp_resistance
+    ))
+    self.amp_fet = self.Block(Fet.NFet(
+      drain_voltage=control_voltage,
+      drain_current=RangeExpr.ZERO,  # effectively no current
+      gate_voltage=(self.control.link().output_thresholds.upper(), self.control.link().voltage.upper())
+    ))
+
+    self.ctl_diode = self.Block(Diode(
+      reverse_voltage=control_voltage,
+      current=RangeExpr.ZERO,  # effectively no current
+      voltage_drop=self.diode_drop,
+      reverse_recovery_time=RangeExpr.ALL
+    ))
+
+    self.btn_diode = self.Block(Diode(
+      reverse_voltage=control_voltage,
+      current=RangeExpr.ZERO,  # effectively no current
+      voltage_drop=self.diode_drop,
+      reverse_recovery_time=RangeExpr.ALL
+    ))
+
+    self.import_kicad(self.file_path("resources", f"{self.__class__.__name__}.kicad_sch"),
+                      conversions={
+                        'pwr_in': VoltageSink(
+                          current_draw=self.pwr_out.link().current_drawn,
+                          voltage_limits=RangeExpr.ALL,
+                        ),
+                        'pwr_out': VoltageSource(
+                          voltage_out=self.pwr_in.link().voltage,
+                          current_limits=RangeExpr.ALL,
+                        ),
+                        'control': DigitalSink(),  # TODO more modeling here?
+                        'gnd': Ground(),
+                        'btn_out': DigitalBidir(
+                          voltage_out=self.gnd.link().voltage - self.diode_drop.upper(),
+                          output_thresholds=(self.gnd.link().voltage.upper(), float('inf')),
+                        ),
+                        'btn_in': DigitalSingleSource(
+                          voltage_out=self.gnd.link().voltage,
+                          output_thresholds=(self.gnd.link().voltage.upper(), float('inf')),
+                          low_signal_driver=True
+                        )
+                      })
+
+
+class SoftPowerSwitch(PowerSwitch, Block):
+  """A software power switch that adds a power button a user can turn on
+  """
+  @init_in_parent
+  def __init__(self, pull_resistance: RangeLike = 10 * kOhm(tol=0.05), amp_resistance: RangeLike = 10 * kOhm(tol=0.05),
+               diode_drop: RangeLike = (0, 0.4) * Volt):
+    super().__init__()
+
+    self.gnd = self.Port(Ground.empty(), [Common])
```

Comment:
Stylistically, consider instantiating the SoftPowerGate and exporting its gnd. Then, create and connect btn later - can be in `contents`

Reply:
applied in 9c3f75d